modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -110,6 +122,654 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred RelSlot to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStatBySlot(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexSlot pred) {
+        Set<RelColumnOrigin> origins = null;
+
+        if (pred instanceof RexLocalRef) {
+            if (rel instanceof ProjectableFilterableTableScan) {
+                origins = Collections.singleton(
+                    ((ProjectableFilterableTableScan)rel).columnOriginsByRelLocalRef(pred.getIndex()));
+            }
+        }
+        else
+            origins = mq.getColumnOrigins(rel, pred.getIndex());
+
+        if (origins == null || origins.isEmpty() || origins.iterator().next().isDerived())
+            return null;
+
+        IgniteTypeFactory typeFactory = Commons.typeFactory(rel);
+
+        List<String> columns = tbl.getRowType(typeFactory).getFieldNames();
+
+        String colName = columns.get(origins.iterator().next().getOriginColumnOrdinal());
+
+        if (QueryUtils.KEY_FIELD_NAME.equals(colName))
+            colName = tbl.descriptor().typeDescription().keyFieldName();
+
+        Statistic stat = tbl.getStatistic();
+
+        if (stat == null)
+            return null;
+
+        return ((IgniteStatisticsImpl)stat).getColumnStatistics(colName);
+    }
+
+    /**
+     * Estimate range selectivity based on predicate.
+     *
+     * @param colStat Column statistics to use.
+     * @param pred  Condition.
+     * @return Selectivity.
+     */
+    private double estimateRangeSelectivity(ColumnStatistics colStat, RexNode pred) {
+        if (pred instanceof RexCall) {
+            RexLiteral literal = getOperand(pred, RexLiteral.class);
+
+            if (literal == null)
+                return 1.;
+
+            BigDecimal val = toComparableValue(literal);
+
+            return estimateSelectivity(colStat, val, pred);
+        }
+
+        return 1.;
+    }
+
+    /**
+     * Estimate range selectivity based on predicate, condition and column statistics.
+     *
+     * @param colStat Column statistics to use.
+     * @param val Condition value.
+     * @param pred Condition.
+     * @return Selectivity.
+     */
+    private double estimateSelectivity(ColumnStatistics colStat, BigDecimal val, RexNode pred) {
+        // Without value or statistics we can only guess.
+        if (val == null)
+            return guessSelectivity(pred);
+
+        SqlOperator op = ((RexCall)pred).op;
+
+        BigDecimal min = toComparableValue(colStat.min());
+        BigDecimal max = toComparableValue(colStat.max());
+        BigDecimal total = (min == null || max == null) ? null : max.subtract(min).abs();
+
+        if (total == null)
+            // No min/max mean that all values are null for coumn.
```

Comment:
`toComparableValue` returns NULL for string and byte array values, causing a string index be chosen over any other. Need to fix this and add an according test for this case.